### ISSUE TITLE
动态图下，在第一次运行后，中断guard 再起新guard， 会出现参数无法更新的现象

### ISSUE THREAD

**Issue by xubenben** [PaddlePaddle/Paddle#20298]
- 版本、环境信息：
    1）PaddlePaddle版本：paddlepaddle==1.5.2
    4）系统环境：Python3.7
- 训练信息
    1）单机
- 问题描述：动态图下，在第一次运行后，中断guard 再起新guard， 会出现参数无法更新的现象

import numpy as np
import paddle
import paddle.fluid as fluid
import paddle.fluid.dygraph as dygraph
place = fluid.CPUPlace()
def reader(num_examples):
    xs = np.random.rand(num_examples, 5).astype("float32")
    ys = np.random.rand(num_examples, 1).astype("float32")
    def __reader__():
        for x, y in zip(xs, ys):
            yield x, y
    return __reader__
def create_model():
    fc = dygraph.FC("FC", 10)
    adam = fluid.optimizer.AdamOptimizer(learning_rate=0.001)
    return fc, adam
def run_model(reader, model, optimizer):
    total_avg_loss = 0
    total_cnt = 0
    reader = paddle.batch(reader, 32)
    for batch in reader():
        x, y = zip(*batch)
        x = np.stack(x, 0)
        y = np.stack(y, 0)
        x = dygraph.to_variable(x)
        y = dygraph.to_variable(y)
        pred = model(x)
        loss = fluid.layers.square_error_cost(pred, y)
        avg_loss = fluid.layers.mean(loss)
        avg_loss.backward()
        optimizer.minimize(avg_loss)
        avg_loss = avg_loss.numpy()[0]
        total_avg_loss += avg_loss
        total_cnt += 1
        model.clear_gradients()
    print(total_avg_loss / total_cnt)
model, optimizer = create_model()
with dygraph.guard(place):
    data_reader = reader(1000)
    for epoch in range(4):
        run_model(data_reader, model, optimizer)
with dygraph.guard(place):
    for epoch in range(10):
        run_model(data_reader, model, optimizer)

在代码最后几行，我先中断当前guard，再起一个新guard，然后参数便无法跟新，这是输出结果：
0.4249080354347825
0.33915749890729785
0.27647794131189585
0.2311921687796712
0.2125754221342504
0.2125754221342504
0.2125754221342504
0.2125754221342504
0.2125754221342504
0.2125754221342504
0.2125754221342504
0.2125754221342504
0.2125754221342504
0.2125754221342504

**Comment by DDDivano** [comment]
中止guard然后另起一个新guard是为了实现是什么功能吗？

**Comment by JiabinYang** [comment]
这里
1. `create model`需要在`guard`中进行，在同一个`guard`中参数才会被`hold`，否则参数会因为离开作用域而被python析构， 而这里能进行更新是因为`FC`比较特殊的会在执行了第一次时才会init，因此他的参数是被第一个`guard`所持有。

2. 不同guard之中的参数如果需要共享，需要通过赋值操作，把第一个`guard`中的参数赋值到第二个`guard`中，并且在更新时传入`optimizer.minimze(param_list=model.parameters())`

**Comment by JiabinYang** [comment]
而你目前的code的话，由于第二个guard中没有`trace`任何参数， 所以没有更新，如果希望更新可以在`run_model` 方法中，将`optimizer.minimize(avg_loss)` 改为`optimizer.minimize(avg_loss， parameter_list=model.parameters())`

**Comment by xubenben** [comment]
赋值操作是具体怎么操作？目前的代码通过设置已经可以了 @JiabinYang
